src/Compilers/Core/Portable/Collections/TopologicalSort.cs

```diff
@@ -24,14 +35,19 @@ internal static class TopologicalSort
         /// </summary>
         /// <typeparam name="TNode">The type of the node</typeparam>
         /// <param name="nodes">Any subset of the nodes that includes all nodes with no predecessors</param>
-        /// <param name="successors">A function mapping a node to its set of successors</param>
+        /// <param name="addSuccessors">A function that adds successor nodes to a provided <see cref="TemporaryArray{TNode}"/>.</param>
         /// <param name="result">A list of all reachable nodes, in which each node always precedes its successors</param>
         /// <returns>true if successful; false if not successful due to cycles in the graph</returns>
-        public static bool TryIterativeSort<TNode>(IEnumerable<TNode> nodes, Func<TNode, ImmutableArray<TNode>> successors, out ImmutableArray<TNode> result)
+        public static bool TryIterativeSort<TNode>(
+            IEnumerable<TNode> nodes,
+            TopologicalSortAddSuccessors<TNode> addSuccessors,
+            out ImmutableArray<TNode> result)
             where TNode : notnull
         {
             // First, count the predecessors of each node
-            PooledDictionary<TNode, int> predecessorCounts = PredecessorCounts(nodes, successors, out ImmutableArray<TNode> allNodes);
+            PooledDictionary<TNode, int> predecessorCounts = PredecessorCounts(nodes, addSuccessors, out ImmutableArray<TNode> allNodes);
+
+            using var successors = TemporaryArray<TNode>.Empty;
```

Comment:
Not necessarily blocking the PR since I don't know how we're going to untangle the problem, but the disposal pattern for `TemporaryArray` was let in unintentionally into the compiler, so I hesitate to pile onto the problem before it can be resolved.

Reply:
> but the disposal pattern for TemporaryArray was let in unintentionally into the compiler

I'm not sure what you mean here :)

This is a Disposable type, and there's nothign wrong with disposable types in the compiler layer (never has been).  We use them all over the place (for example, the parser) for very clean code that is safe to making mistakes (the sorts both you and i have found in recent PRs).

The Dispose concern existed for types that are also manually *Freeable*.  Because there was a concern of someone both disposing and freeing an item.  TempArray exposes no free mechanism *intentionally*.  So you can't accidentally fall into that path.  This is a core part of the design precisely so that it can fit in both IDE and compiler and be used properly with `using` without issue in either layer.

Reply:
This is also why the PR that adds *Free* to TempArray is problematic.  It actively moves TempArray into having a duality that will actively make it easier to make mistakes with.  We do *not* want such a Duality, and instead want this to behave like normal Disposable types (including the other ones used in the compiler) that only have one clear form that is basically impossible to get wrong :)

Reply:
It feels unfortunate that in order to get "do something whenever we exit this block through a `return`" in C# we also have to opt-in to the overhead of "do something whenever we exit this block through a `throw`".

Reply:
Note: overhead appears to be... nothing/near-nothing? :)  As an example, tehse sorts of patterns are used extensively in the parser, especially in hotpaths around speculation.  I've tested with/without the using, and found the perf effectively identical.  Maybe this was a concern (or perhaps in hte netfx days?).  It doesn't seem to matter today.  Or if it does exist, it's so far dwarfed by the other work we do around this.

In cases like this, for example, we get such nice wins with data locality just using the stack and not producing garbage, that a simple try/finally seems irrelevant in comparison.